text/3446-store.md

```diff
@@ -0,0 +1,1246 @@
+        the maximum capacity once and for all._
+
+
+##  Guarantees, or absence thereof
+
+The `Store` API is minimalist, providing a minimum of guarantees.
+
+Beyond being untyped and unchecked, there are also a few oddities, compared to the `Allocator` API:
+
+-   By default, using `Store::allocate` or `Store::allocate_zeroed` invalidates all existing handles. This oddity
+    stems from the requirement of minimizing the state to be stored in collections using a single allocation at a time
+    such as `Box`, `Vec`, `VecDeque`, ...
+-   By default, calling any method -- including `resolve` -- invalidates all resolved pointers[^1]. This oddity stems
+    from the desire to leave the API flexible enough to allow caching stores, single-allocation stores, or copying
+    stores.
+-   By default, moving `Store` invalidates all resolved pointers. This oddity stems from the fact that when using an
+    in-line store the pointers point within the block of memory covered by the store, and thus after it moved, are left
+    pointing into the void.
+
+When the above should be guaranteed, extra marker traits can be implemented to provide compile-time checking that these
+properties hold, which in turn allows the final user to safely mix and match collection and store, relying on compiler
+errors to prevent erroneous couplings.
+
+[^1]: With the exception, in the case of a call to `resolve`, of any pointer derived from a copy of the handle argument.
+
+
+##  Mutable Store: allocation.
+
+A previous incarnation of the API borrowed the store mutably to allocate, deallocate, grow, or shrink.
+
+This is tempting, as after all it is likely something within the store will need to be mutated.
+
+There is, however, a very good reason for `Allocator` to use a shared reference: concurrent uses. In a concurrent
+context, a `Sync` allocator is shared between threads.
+
+Suggestions were made to use `for<'a> &'a S: Store` in such cases, however, to the best of my knowledge, it is not
+possible today to refer to the associated type of such a bound. That is, it is not possible to declare a field as
+`handle: <for<'a> &'a S: Store>::Handle`, which greatly complicates things...
+
+
+##  Mutable Store: resolution.
+
+A previous incarnation of the API borrowed the store mutably to resolve mutable handles, ie it offered both a `resolve`
+and a `resolve_mut` methods.
+
+While this would, theoretically, allow implementing a store with no interior mutability, it unfortunately seems to run
+afoul of aliasing when obtaining multiple mutable references to distinct elements with overlapping lifetimes -- such
+as when retaining multiple mutable references yielded by an iterator.
+
+The problem is illustrated in [3446-store/aliasing-and-mutability.rs] on which Stacked Borrows chokes. While Tree
+Borrows _does_ accept the program as valid, it is not clear whether LLVM `noalias` would be compatible in such a case
+or not, now and in the future.
```

Comment:
For all we know, if the code is accepted by Tree Borrows, then it is compatible with LLVM `noalias`. Everything else would be a serious bug in Tree Borrows.